### DISCUSSION TITLE
Run multiple instances of FastAPI with SQLAlchemy using a centralized connection pooling mechanism

### DISCUSSION THREAD

**Discussion by thesubuck** [fastapi/fastapi#9740] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# This is a module where the ORM-SQL Alchemy is configured with connection pool setup
import os
import logging
from dotenv import load_dotenv
from sqlalchemy import create_engine, inspect, text
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

from .query_profiler import SqlProfiler

logger = logging.getLogger(__name__)
load_dotenv()
SqlProfiler()

SQLALCHEMY_DATABASE_URL = os.getenv("SQLALCHEMY_DATABASE_URL")

SQLALCHEMY_ENGINE_OPTIONS = {
    "pool_pre_ping": True,
    "pool_size": 8,
    "max_overflow": 4,
    "pool_timeout": 10,
    "pool_recycle": 300,
}
engine = create_engine(SQLALCHEMY_DATABASE_URL, **SQLALCHEMY_ENGINE_OPTIONS)

SessionLocal = sessionmaker(autocommit=False, autoflush=True, bind=engine)

"""
Now we will use the function declarative_base() that returns a class.
Later we will inherit from this class to create each of the database orm_models or classes (the ORM orm_models).
"""
Base = declarative_base()


def get_db():
    db = SessionLocal()
    logger.info(f"Current Connection Pool Number: {engine.pool.checkedout()}")
    try:
        yield db
    finally:
        db.close()


# This method "get_db()" is used in routers to access the orm session for CRUD operations

# How the Fast API app is run:
uvicorn app:app --host 0.0.0.0 --port 8080 --workers 4
# So there will be 4 instances of Fastapi app running
```

### Context:
I use FastAPI with SQL Alchemy .
SQL Alchemy has connection pooling enabled
The ORM packages interacts with MySQL database with supports 60 database connections
I have setup the pooling parameters such that the sum of all connection pools will be less than the connection limit.
If you see the code, I am using 8+4=12 as total pool size for one FastAPI instance and I have four, which is 4*12 = 48 .

### Problem
One of the FastAPI instance exhausted the pool limit (12) and posted an error stating the same.
However, the other instances still did not exhaust the pool limit and still responding to requests at that time.

### Question
**How to implement a centralized connection pool across multiple instances of Fast API so that the pool is shared across the instances and used intelligently?**

### Operating System

Linux

### FastAPI Version

0.92.0

### Python Version

Python 3.10.4

### Additional Context

SQL Alchemy version : 1.4.47

**Comment by jgould22** [comment, marked as answer]
It would be quiet costly and finicky to share a connection pool across python processes like this so it is not generally done in the application itself.

An option you could look into is placing a connection pooler in front of MySQL (with PostgreSQL this is often done with PGBouncer) and it appears there are some similar options in the MySQL space like Proxysql (though I have not used it so your mileage may vary).

Regardless the approach would be similar no matter what you use. Have a proxy that sits between your application and your database that will hold a connection pool and manage all the state that is required with the database connections and let then let applications connect to that.

All of this assumes that you cant just bump the connection numbers across the board but that depends on what type of queries and what hardware your database is on.

**Reply by thesubuck** [reply]
Thank you Jordan Gould for taking time and writing this . I appreciate it. I am going to explore your approach of DB <--> Proxy <--> API and try to solve this problem. You assumption of bumping up the connection numbers are not an option is correct.

**Reply by thesubuck** [reply]
@jgould22 After going though the RDS connection usage with pooling and without pooling, it is evident that "using sql alchemy without connection pooling" will solve my connection exhaustion problem. That being said, I am going to accept your solution for centralized connection pooling and closing this thread. Thank you once again.
